src/measurement/vmsa.rs

```diff
@@ -0,0 +1,369 @@
+    Sev,
+    /// SEV-ES
+    SevEs,
+    /// SEV-SNP
+    SevSnp,
+}
+
+impl FromStr for SevMode {
+    type Err = MeasurementError;
+
+    fn from_str(s: &str) -> Result<Self, MeasurementError> {
+        match s.to_lowercase().as_str() {
+            "sev" => Ok(SevMode::Sev),
+            "sev-es" | "seves" => Ok(SevMode::SevEs),
+            "sev-snp" | "sevsnp" => Ok(SevMode::SevSnp),
+            _ => Err(MeasurementError::InvalidSevModeError(s.to_string())),
+        }
+    }
+}
+
+/// VmmTypes
+#[derive(Clone, Copy, PartialEq)]
+pub enum VMMType {
+    /// QEMU
+    QEMU = 1,
+    /// EC2
+    EC2 = 2,
+    /// KRUN
+    KRUN = 3,
+}
+
+impl TryFrom<u8> for VMMType {
+    type Error = MeasurementError;
+
+    fn try_from(value: u8) -> Result<Self, MeasurementError> {
+        match value {
+            1 => Ok(VMMType::QEMU),
+            2 => Ok(VMMType::EC2),
+            3 => Ok(VMMType::KRUN),
+            _ => Err(MeasurementError::InvalidVmmError(value.to_string())),
+        }
+    }
+}
+
+impl fmt::Debug for VMMType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            VMMType::QEMU => write!(f, "qemu"),
+            VMMType::EC2 => write!(f, "ec2"),
+            VMMType::KRUN => write!(f, "krun"),
+        }
+    }
+}
+
+/// Virtual Machine Control Block
+/// The layout of a VMCB struct is documented in Table B-1 of the
+/// AMD64 Architecture Programmer’s Manual, Volume 2: System Programming
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Serialize, Default)]
+struct VmcbSeg {
+    /// Segment selector: documented in Figure 4-3 of the
+    /// AMD64 Architecture Programmer’s Manual, Volume 2: System Programming
+    selector: u16,
+
+    /// Segment attributes.
+    attrib: u16,
+
+    /// Segment limit: used in comparisons with pointer offsets to prevent
+    /// segment limit violations.
+    limit: u32,
+
+    /// Segment base address.
+    base: u64,
+}
+
+impl VmcbSeg {
+    fn new(selector: u16, attrib: u16, limit: u32, base: u64) -> Self {
+        Self {
+            selector,
+            attrib,
+            limit,
+            base,
+        }
+    }
+}
+
+/// Large array structure to serialize and default arrays larger than 32 bytes.
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
+#[repr(C)]
+pub(crate) struct LargeArray<T, const N: usize>(#[serde(with = "BigArray")] [T; N])
+where
+    T: for<'a> Deserialize<'a> + Serialize;
+
+impl<T, const N: usize> Default for LargeArray<T, N>
+where
+    T: std::marker::Copy + std::default::Default + for<'a> Deserialize<'a> + Serialize,
+{
+    fn default() -> Self {
+        Self([T::default(); N])
+    }
+}
+
+/// SEV-ES VMSA page
+/// The names of the fields are taken from struct sev_es_work_area in the linux kernel:
+/// https://github.com/AMDESE/linux/blob/sev-snp-v12/arch/x86/include/asm/svm.h#L318
+/// (following the definitions in AMD APM Vol 2 Table B-4)
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Serialize, Default)]
```

Comment:
Do we need to derive deserialize?
```suggestion
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Default)]
```

Reply:
At no point do I deserialize this struct, but I guess it won't hurt to add it.